node/core/provisioner/src/lib.rs

```diff
@@ -712,7 +714,20 @@ async fn select_disputes(
 	// In case of an overload condition, we limit ourselves to active disputes, and fill up to the
 	// upper bound of disputes to pass to wasm `fn create_inherent_data`.
 	// If the active ones are already exceeding the bounds, randomly select a subset.
-	let recent = request_disputes(sender, RequestType::Recent).await;
+	let mut recent = request_disputes(sender, RequestType::Recent).await;
+
+	// On chain disputes are fetched for the runtime. We don't want to include disputes the runtime already knows
+	// about in the inherent data.
+	// On error - an empty set is generated which doesn't affect the logic of the function.
+	// This is a staging feature, so if it is not enabled - create an empty HashSet by default. If it is
+	// enabled - the HashSet will be overriden with the on-chain data.
+	let onchain: HashSet<(SessionIndex, CandidateHash)> = HashSet::new();
+	#[cfg(feature = "improved-onchain-disputes-import")]
```

Comment:
Why is the feature needed ?

Reply:
It uses a staging API call so to be sure this doesn't end up on production networks the client code is behind a feature flag.

Reply:
Okay. Let's have one feature for all staging client code - `staging-client` (better names a are welcome)?
It will be used to enable/disable the client code using the staging runtime api (which is controlled with `vstaging` feature).

We can't use `vstaging` for both, because its purpose is to enable the runtime api itself on different networks, while the client code is common everywhere.